### PR TITLE
Upgrade to modern spark (3.4)

### PR DESCRIPTION
**Pull request by holdenk**
Hey y'all poked around in the project, I don't know if this is still being actively developed but if folks are interested upgraded it to Spark 3.4, I've only done the tests with the built in test suite and it's my first time poking around in the project so you know could be some mistakes.

### CODE REVIEW

**Comment by colindean** [comment]
I _really_ appreciate the contribution — first outside contribution in some time! For our purposes, DV _must_ continue working with Spark 2.3 and JDK8, so we can't bump versions as you have proposed.

Can you think of a way to test/qualify this against both Spark 2.3 and 3.4? As a potential starting point, I started on [crossbuilding](https://github.com/target/data-validator/compare/master...gh-84-crossbuild) last year but haven't had the bandwidth to complete the work.

**Comment by holdenk** [comment]
So one lazy option would be to make a spark 3 and a spark 2 branch and publish a "-3" version of the artifact rather than a "proper" cross build.

**Comment by c-horn** [comment]
I could potentially take a look at finishing the cross-build work @colindean 
Are there still users of this internally at Target who can't switch to Spark 3.X + docker?

Also many thanks @holdenk for keeping this project / others in the spark ecosystem in mind.

**Comment by colindean** [comment]
> So one lazy option would be to make a spark 3 and a spark 2 branch and publish a "-3" version of the artifact rather than a "proper" cross build.

That's a good idea, but it introduces some operational complexity to be recalled at release time. I prefer the technical complexity of an sbt crossbuild, especially if…

> I could potentially take a look at finishing the cross-build work @colindean

> So I was thinking the main "annoying" thing in a cross build for this is probably the tests and maybe the quoted identifiers. We could probably make it cross build easily, although my default is lazy (branches).

❤️ Yes that would be awesome. I don't remember what was going awry when I left off with it. I welcome the contribution, including pairing if either of you'd like (or you could pair the two of you if you'd like).

> Are there still users of this internally at Target who can't switch to Spark 3.X + docker?

I've not seen pressure from DV users to support Spark 3, but most DV users who have identified themselves are internal to Target and stuck on our Spark 2.3 cluster while we await the production launch of our internal platform enabling use of Spark 3 + Docker.

> Also many thanks @holdenk for keeping this project / others in the spark ecosystem in mind.

Yes, indeed! I'm very happy to see this and your continued work in the space, and to see another former IBMer around!

**Comment by holdenk** [comment]
So my other project where I do more extensive cross-building is https://github.com/holdenk/spark-testing-base/blob/main/build.sbt , it puts the actual Spark Versions outside of sbt (passed in as a parameter) and does "classic" sbt cross-build for the scala versions. What do folks think of something like that?

Also thanks for being so friendly y'all :)

**Comment by holdenk** [comment]
Also, slightly off topic, (in the context of having a large pre-Spark 3 codebase) my self and some of my coworkers have been working on some tooling at https://github.com/holdenk/spark-upgrade that might be useful once you'all have Spark 3 ready for use :)

**Comment by holdenk** [comment]
A bit of work to see if folks agree with the direction, if so I'll update the tests and such :)

**Comment by holdenk** [comment]
Hey just following up @c-horn do you have some bandwidth to review this?

File: build.sbt [hunk 1]
```diff
@@ -35,18 +66,26 @@ libraryDependencies ++= Seq(
...
 )
 
 Test / fork := true
-javaOptions ++= Seq("-Xms512M", "-Xmx2048M", "-XX:+CMSClassUnloadingEnabled")
+javaOptions ++= (if (sparkVersion.value > "3.0") {
+  Seq("-Xms4048M", "-Xmx4048M",
+    "-Dio.netty.tryReflectionSetAccessible=true",
+    "--add-opens=java.base/java.lang=ALL-UNNAMED",
+    "--add-opens=java.base/java.io=ALL-UNNAMED",
+    "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED")
```
[the remark this answers is not in the document]
Comment (colindean) [reply build.sbt:85]: Mostly because https://github.com/target/data-validator/pull/166/checks#step:5:58

**Comment by colindean** [comment]
It's been a while but I'm hoping to revisit this mid-February. I shook out some backed-up update PRs and will take a close look at this when I return from FOSDEM — @holdenk, I noted that you're speaking there and I hope to attend your talk and meet you!

**Comment by holdenk** [comment]
Sounds awesome, maybe we can grab some coffee:)

**Comment by colindean** [comment]
Good news! The requirement to support Spark 2 is now gone as my org is moving to Spark 3 by the end of May. I'm going to return to this in the coming days/weeks as I've gotten a few requests for it.

**Comment by colindean** [comment]
We're going into Spark 3 at v3.5.1 and still supporting Scala 2.12 and 2.13.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -23,5 +23,7 @@ jobs:
       - uses: coursier/setup-action@v1
         with:
           jvm: adopt:1.8
-      - name: Build, test, and package project
-        run: bin/sbt clean compile test package makePom
+      - name: Build, test, and package project on Spark 3.3
+        run: bin/sbt clean compile test package makePom -DsparkVersion=3.3.1
```
Comment (colindean) [inline .github/workflows/ci.yaml:27]: **suggestion:** 
```suggestion
        run: bin/sbt clean compile test package makePom -DsparkVersion=3.5.1
```

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -25,3 +25,5 @@ jobs:
       # uses sbt-github-packages, see build.sbt
       - name: Publish with SBT
         run: bin/sbt publish
+      - name: Publish with SBT
+        run: bin/sbt publish -DsparkVersion=3.3.1
```
Comment (colindean) [inline .github/workflows/release.yaml:29]: **suggestion:** 
```suggestion
        run: bin/sbt publish -DsparkVersion=3.5.1
```

File: README.md [hunk 4]
```diff
@@ -65,6 +65,8 @@ Usage: data-validator [options]
   --help                   Show this help message and exit.
 ```
 
+If you want to build with a modern version of java set the "MODERN_JAVA" enviornment variable.
```
Comment (colindean) [inline README.md:68]: **suggestion:** 
```suggestion
If you want to build with a modern version of Java, set the "MODERN_JAVA" environment variable.
```

Also, @holdenk, can you say what constitutes a modern version? I reckon it'd be 17+?

**Comment by colindean** [comment]
#186 sets some new baseline. I think I want to get this merged (probably with the suggestions to move to Spark 3.5.1 merged) and then continue from there.

File: build.sbt (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaspal1404) [reply build.sbt:86]: @holdenk Where are we setting the Spark version in this new change ?

[the remark this answers is not in the document]
Comment (jaspal1404) [reply build.sbt:86]: Getting this error when trying to run the current build with our pipeline validation step on Argus - java.lang.NoSuchMethodError: org.yaml.snakeyaml.constructor.SafeConstructor: method 'void <init>()' not found

**Comment by jaspal1404** [comment]
@colindean @holdenk - Here is the change that is required in build.sbt to make the DV work with Spark version 3.x. I tested it and its working:

current circe-yaml version : 0.14.2
new circe-yaml version : 1.15.0

Change line number 36 in build.sbt from 0.14.2 to 1.15.0

File: build.sbt [hunk 5]
```diff
@@ -1,11 +1,42 @@
...
-scalaVersion := "2.11.12"
+val sparkVersion = settingKey[String]("Spark version")
 
-val sparkVersion = "2.3.4"
+sparkVersion := System.getProperty("sparkVersion", "2.3.4")
 
-val circeVersion = "0.11.2"
+scalaVersion := {
+  if (sparkVersion.value > "3.0") {
+    "2.12.17"
```
Comment (colindean) [inline build.sbt:10]: **suggestion:** 
```suggestion
    "2.12.19"
```

File: build.sbt [hunk 6]
```diff
@@ -1,11 +1,42 @@
...
+sparkValidationVersion := "0.15.0"
+
+version := sparkVersion.value + "_" + sparkValidationVersion.value
+
+val circeVersion = settingKey[String]("Circe version")
+val circeYamlVersion = settingKey[String]("Circe YAML version")
+
+circeVersion := {
+  if (sparkVersion.value > "3.0") {
+    "0.14.2"
```
Comment (colindean) [inline build.sbt:27]: **suggestion:** 
```suggestion
    "0.14.6"
```

File: build.sbt [hunk 7]
```diff
@@ -1,11 +1,42 @@
...
+  } else {
+    "0.11.2"
+  }
+}
+
+circeYamlVersion := {
+  if (sparkVersion.value > "3.0") {
+    "0.14.2"
```
Comment (colindean) [inline build.sbt:35]: **suggestion:** 
```suggestion
    "0.14.6"
```

File: README.md (hunk 4 above)
Comment (colindean) [reply README.md:68]: Based on some searching around, the options added below appear to be available in JDK11+, so I think "modern" means JDK 11+… so I'm going to make it say that for now.

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -23,5 +23,7 @@ jobs:
       - uses: coursier/setup-action@v1
         with:
           jvm: adopt:1.8
-      - name: Build, test, and package project
-        run: bin/sbt clean compile test package makePom
+      - name: Build, test, and package project on Spark 3.3
```
Comment (colindean) [inline .github/workflows/ci.yaml:26]: **suggestion:** 
```suggestion
      - name: Build, test, and package project on Spark 3.5
```

File: build.sbt (hunk 1 above)
[the remark this answers is not in the document]
Comment (colindean) [reply build.sbt:86]:
```suggestion
    // these were added in JDK 11 and newer, apparently.
    "-Dio.netty.tryReflectionSetAccessible=true",
    "--add-opens=java.base/java.lang=ALL-UNNAMED",
    "--add-opens=java.base/java.io=ALL-UNNAMED",
    "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED")
```

File: build.sbt [hunk 9]
```diff
@@ -1,11 +1,42 @@
...
+  if (sparkVersion.value > "3.0") {
+    "0.14.6"
+  } else {
+    "0.11.2"
+  }
+}
+
+circeYamlVersion := {
+  if (sparkVersion.value > "3.0") {
+    "0.14.6"
```
Comment (colindean) [inline build.sbt:35]: **nitpick:** 
```suggestion
    "0.15.1"
```

need a version that exists

**Comment by colindean** [comment]
Thank you so much, @holdenk, for submitting this. I'm sorry it took a while for priorities to bubble up. I'm glad that we were able to merge it eventually.

**Review by jaspal1404 — COMMENTED** [review]
looks good!

**Pull request merged**